### PR TITLE
HW49-solution

### PR DESCRIPTION
**Pull request by OleksiiKolinko**
test

### CODE REVIEW

File: src/test/java/mate/academy/controller/BookControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+import org.springframework.test.context.jdbc.Sql;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
+import org.testcontainers.shaded.org.apache.commons.lang3.builder.EqualsBuilder;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+public class BookControllerTest {
+    protected static MockMvc mockMvc;
```
Comment (Sanderfuck) [inline src/test/java/mate/academy/controller/BookControllerTest.java]: Why protected?

File: src/test/java/mate/academy/controller/BookControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    @SneakyThrows
+    static void teardown(DataSource dataSource) {
+        try (Connection connection = dataSource.getConnection()) {
+            connection.setAutoCommit(true);
+            ScriptUtils.executeSqlScript(connection,
+                    new ClassPathResource("database/books/remove-book-from-books-table.sql"));
+        }
+    }
+
+    private List<BookDto> getBookDtos() {
```
Comment (Sanderfuck) [inline src/test/java/mate/academy/controller/BookControllerTest.java]: Private methods place after public

File: src/test/java/mate/academy/controller/BookControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        bookDto3.setDescription("Description3");
+        bookDto3.setCoverImage("Cover Image3");
+        Set<CategoryDto> categoryDtos3 = new HashSet<>();
+        categoryDtos3.add(categoryDto2);
+        bookDto3.setCategories(categoryDtos3);
+        List<BookDto> expected = new ArrayList<>();
+        expected.add(bookDto1);
+        expected.add(bookDto2);
+        expected.add(bookDto3);
+        return expected;
```
Comment (Sanderfuck) [inline src/test/java/mate/academy/controller/BookControllerTest.java]: Create separate templates

File: src/test/java/mate/academy/controller/BookControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        categoryDtos1.add(categoryDto1);
+        bookRequestDto.setCategories(categoryDtos1);
+        BookDto expected = new BookDto();
+        expected.setTitle(bookRequestDto.getTitle());
+        expected.setAuthor(bookRequestDto.getAuthor());
+        expected.setIsbn(bookRequestDto.getIsbn());
+        expected.setPrice(bookRequestDto.getPrice());
+        expected.setDescription(bookRequestDto.getDescription());
+        expected.setCoverImage(bookRequestDto.getCoverImage());
+        expected.setCategories(categoryDtos1);
```
Comment (Sanderfuck) [inline src/test/java/mate/academy/controller/BookControllerTest.java]: Create templates for clear code

File: src/test/java/mate/academy/controller/CategoryControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        teardown(dataSource);
+        try (Connection connection = dataSource.getConnection()) {
+            connection.setAutoCommit(true);
+            ScriptUtils.executeSqlScript(connection,
+                    new ClassPathResource("database/categories/add-three-categories.sql"));
+        }
+    }
+
+    @SneakyThrows
+    static void teardown(DataSource dataSource) {
```
Comment (Sanderfuck) [inline src/test/java/mate/academy/controller/CategoryControllerTest.java]: package private modifier for this method?

File: src/test/java/mate/academy/service/BookServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        BOOK_DTO.setCoverImage(BOOK.getCoverImage());
+        BOOK_DTO.setCategories(CATEGORY_DTOS);
+        CREATE_BOOK_DTO = new CreateBookRequestDto();
+        CREATE_BOOK_DTO.setTitle(BOOK.getTitle());
+        CREATE_BOOK_DTO.setAuthor(BOOK.getAuthor());
+        CREATE_BOOK_DTO.setIsbn(BOOK.getIsbn());
+        CREATE_BOOK_DTO.setPrice(BOOK.getPrice());
+        CREATE_BOOK_DTO.setDescription(BOOK.getDescription());
+        CREATE_BOOK_DTO.setCoverImage(BOOK.getCoverImage());
+        CREATE_BOOK_DTO.setCategories(CATEGORY_DTOS);
```
Comment (Sanderfuck) [inline src/test/java/mate/academy/service/BookServiceTest.java]: Create templates

File: src/test/java/mate/academy/service/BookServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+    @Test
+    @DisplayName("Verify the correct book was return when book exists")
+    public void findById_WithValidBookId_ShouldReturnValidBook() {
+        when(bookRepository.findById(anyLong())).thenReturn(Optional.of(BOOK));
+        when(bookMapper.toDto(any(Book.class))).thenReturn(BOOK_DTO);
+        BookDto actual = bookService.findById(BOOK.getId());
+        assertEquals(BOOK_DTO, actual);
+        verify(bookRepository, times(1)).findById(anyLong());
+        verifyNoMoreInteractions(bookRepository);
```
Comment (Sanderfuck) [inline src/test/java/mate/academy/service/BookServiceTest.java]: For repository you can use testcontainer instead of mock

**Review by Sanderfuck — CHANGES_REQUESTED** [review]

**Review by koval-kn — APPROVED** [review]

**Pull request merged**